src/core_plugins/region_map/public/choropleth_layer.js

```diff
@@ -0,0 +1,287 @@
+      this._legendQuantizer = d3.scale.quantize().domain(quantizeDomain).range(this._legendColors);
+    }
+    this.emit('styleChanged', {
+      mismatches: styler.getMismatches()
+    });
+  }
+
+  getMetrics() {
+    return this._metrics;
+  }
+
+  getMetricsAgg() {
+    return this._metricsAgg;
+  }
+
+  getUrl() {
+    return this._geojsonUrl;
+  }
+
+  setTooltipFormatter(tooltipFormatter, metricsAgg, fieldName) {
+    this._tooltipFormatter = (geojsonFeature) => {
+      if (!this._metrics) {
+        return '';
+      }
+      const match = this._metrics.find((bucket) => {
+        return bucket.term === geojsonFeature.properties[this._joinField];
+      });
+      return tooltipFormatter(metricsAgg, match, fieldName);
+    };
+  }
+
+  setJoinField(joinfield) {
+    if (joinfield === this._joinField) {
+      return;
+    }
+    this._joinField = joinfield;
+    this._setStyle();
+  }
+
+
+  setMetrics(metrics, metricsAgg) {
+    this._metrics = metrics;
+    this._metricsAgg = metricsAgg;
+    this._valueFormatter = this._metricsAgg.fieldFormatter();
+    this._setStyle();
+  }
+
+  setColorRamp(colorRamp) {
+    if (_.isEqual(colorRamp, this._colorRamp)) {
+      return;
+    }
+    this._colorRamp = colorRamp;
+    this._setStyle();
+  }
+
+  equalsGeoJsonUrl(geojsonUrl) {
+    return this._geojsonUrl === geojsonUrl;
+  }
+
+  appendLegendContents(jqueryDiv) {
+
+
+    if (!this._legendColors || !this._legendQuantizer || !this._metricsAgg) {
+      return;
+    }
+
+    const titleText = this._metricsAgg.makeLabel();
+    const $title = $('<div>').addClass('tilemap-legend-title').text(titleText);
+    jqueryDiv.append($title);
+
+    this._legendColors.forEach((color) => {
+
+      const labelText = this._legendQuantizer
+        .invertExtent(color)
+        .map(this._valueFormatter)
+        .join(' – ');
+
+      const label = $('<div>');
+      const icon = $('<i>').css({
+        background: color,
+        'border-color': makeColorDarker(color)
+      });
+
+      const text = $('<span>').text(labelText);
+      label.append(icon);
+      label.append(text);
+
+      jqueryDiv.append(label);
+    });
+
+  }
+
+}
+
+
+function makeColorDarker(color) {
+  const amount = 1.3;//magic number, carry over from earlier
+  return d3.hcl(color).darker(amount).toString();
+}
+
+
+function getMinMax(data) {
+  let min = data[0].value;
+  let max = data[0].value;
+  for (let i = 1; i < data.length; i += 1) {
+    min = Math.min(data[i].value, min);
+    max = Math.max(data[i].value, max);
+  }
+  return { min, max };
+}
+
+
+function makeChoroplethStyler(data, colorramp, joinField) {
+
+
+  if (data.length === 0) {
+    return {
+      getLeafletStyleFunction: function () {
+        return emptyStyle();
+      },
+      getMismatches: function () {
+        return [];
+      }
+    };
+  }
+
+  const { min, max } = getMinMax(data);
+  data = data.slice();
```

Comment:
yes, that order is significant. the getMisMatches is just for the client to be able to read out the errors of the inner-join (ie. the style function).